File: mcp_server/explain.py
"""Turn a result's ``debug`` payload into one sentence a model can reason about.

``debug`` is a DRF ``DictField``, so the published schema types it as an open
object - which is true, and means there is no contract to rely on. It is
therefore narrowed here at runtime, and the narrowing **fails closed**: an
unrecognised shape produces "no ranking explanation available" rather than a
confident sentence assembled from missing keys.

That bias is deliberate. This text exists so a caller can judge how much to
trust a hit. An explanation that is quietly wrong is worse than none at all,
because it is indistinguishable from one that is right.
"""

from typing import Any

# Reciprocal rank fusion contributes 1 / (k + rank) per retriever that found a
# document. Reported so the arithmetic can be checked rather than believed.
_UNKNOWN = "No ranking explanation available for this result."


def _as_number(value: Any) -> float | None:
    # bool is an int subclass, and True would format as 1.0000 - which looks
    # like a score rather than a type error.
    if isinstance(value, bool):
        return None
    if isinstance(value, (int, float)):
        return float(value)
    return None


def _describe_run(name: str, run: Any) -> str | None:
    if not isinstance(run, dict):
        return None
    rank = _as_number(run.get("rank"))
    score = _as_number(run.get("score"))
    contribution = _as_number(run.get("contribution"))
    if rank is None:
        return None

    parts = [f"{name} ranked it #{rank:g}"]
    if score is not None:
        parts.append(f"score {score:.4f}")
    if contribution is not None:
        parts.append(f"contributing {contribution:.5f}")
    return ", ".join(parts)


def explain(debug: Any) -> str:
    """Describe why a paper is at its position, or say that we cannot."""
    if not isinstance(debug, dict):
        return _UNKNOWN

    runs = debug.get("runs")
    if isinstance(runs, dict) and runs:
        described = [
            text
            for name, run in runs.items()
            if (text := _describe_run(str(name), run)) is not None
        ]
        if not described:
            return _UNKNOWN

        sentence = "Fused from " + "; ".join(described) + "."

        rrf_k = _as_number(debug.get("rrf_k"))
        if rrf_k is not None:
            sentence += f" Reciprocal rank fusion with k={rrf_k:g}."

        if len(described) == 1:
            # The single most useful thing this panel can say. One retriever
            # finding a paper the other missed is the entire argument for
            # fusing them, and also the usual explanation for a result that
            # looks wrong.
            sentence += " Only one retriever found this paper; the other missed it."

        degraded = debug.get("degraded")
        if isinstance(degraded, list) and degraded:
            names = ", ".join(str(d) for d in degraded)
            sentence += (
                f" WARNING: {names} failed during this search, so these results"
                " do not reflect the full configured system."
            )
        return sentence

    retriever = debug.get("retriever")
    if isinstance(retriever, str) and retriever:
        similarity = _as_number(debug.get("similarity"))
        bm25_score = _as_number(debug.get("bm25_score"))
        if similarity is not None:
            return f"Retrieved by {retriever}, cosine similarity {similarity:.4f}."
        if bm25_score is not None:
            return f"Retrieved by {retriever}, BM25 score {bm25_score:.4f}."
        return f"Retrieved by {retriever}."

    # Reached when the payload is a dict but matches neither known shape,
    # which is what a serializer change would look like from here.
    return _UNKNOWN

File: frontend/src/api/explain.ts
/**
 * Turning the untyped `debug` blob into something the UI can render.
 *
 * `debug` is a `DictField` on the backend, so the generated schema types it as
 * `Record<string, unknown>` - which is *correct*. The OpenAPI document genuinely
 * does not describe its contents, and pretending otherwise with a hand-written
 * interface would be inventing a contract the server never agreed to.
 *
 * So it is narrowed here, once, at runtime, with guards that fail closed. A
 * shape we do not recognise renders as "no explanation available" rather than
 * as a plausible-looking panel built from undefined values. An explanation that
 * is quietly wrong is worse than no explanation, because the entire point of
 * this panel is to be the thing you trust when a ranking looks odd.
 */

/** One retriever's contribution to a fused result. */
export interface RunContribution {
  retriever: string;
  rank: number;
  score: number;
  contribution: number;
}

export type Explanation =
  | {
      kind: "fused";
      rrfK: number;
      runs: RunContribution[];
      /** Retrievers that failed on this request, if any. */
      degraded: string[];
    }
  | {
      kind: "single";
      retriever: string;
      score: number;
      /** Which chunk of the paper matched. >0 means a split abstract. */
      chunkIndex: number | null;
      embeddingModel: string | null;
    }
  | { kind: "unknown" };

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function asNumber(value: unknown): number | null {
  return typeof value === "number" && Number.isFinite(value) ? value : null;
}

function asStringArray(value: unknown): string[] {
  if (!Array.isArray(value)) return [];
  return value.filter((item): item is string => typeof item === "string");
}

export function explain(debug: Record<string, unknown>): Explanation {
  const rrfK = asNumber(debug["rrf_k"]);
  const runsRaw = debug["runs"];

  if (rrfK !== null && isRecord(runsRaw)) {
    const runs: RunContribution[] = [];
    for (const [retriever, detail] of Object.entries(runsRaw)) {
      if (!isRecord(detail)) continue;
      const rank = asNumber(detail["rank"]);
      const score = asNumber(detail["score"]);
      const contribution = asNumber(detail["contribution"]);
      if (rank === null || score === null || contribution === null) continue;
      runs.push({ retriever, rank, score, contribution });
    }
    if (runs.length === 0) return { kind: "unknown" };

    // Largest contribution first: the retriever that actually decided this
    // ranking should be the first thing read, not an alphabetical accident.
    runs.sort((a, b) => b.contribution - a.contribution);
    return {
      kind: "fused",
      rrfK,
      runs,
      degraded: asStringArray(debug["degraded"]),
    };
  }

  const retriever = debug["retriever"];
  if (typeof retriever === "string") {
    // bm25 reports `bm25_score`, dense reports `similarity`. Neither is a
    // probability and they are not comparable to each other, which is exactly
    // why RRF fuses ranks rather than these numbers.
    const score = asNumber(debug["bm25_score"]) ?? asNumber(debug["similarity"]);
    if (score === null) return { kind: "unknown" };
    const embeddingModel = debug["embedding_model"];
    return {
      kind: "single",
      retriever,
      score,
      chunkIndex: asNumber(debug["chunk_index"]),
      embeddingModel: typeof embeddingModel === "string" ? embeddingModel : null,
    };
  }

  return { kind: "unknown" };
}

/** Retrievers that were asked but returned nothing for this paper. */
export function missingFrom(explanation: Explanation, allRetrievers: string[]): string[] {
  if (explanation.kind !== "fused") return [];
  const found = new Set(explanation.runs.map((run) => run.retriever));
  return allRetrievers.filter((name) => !found.has(name));
}
